src/ext/detective.lisp

```diff
@@ -0,0 +1,282 @@
+
+(defclass reference ()
+  ((name :type string 
+         :initarg :reference-name
+         :reader reference-name)
+
+   (point :initarg :reference-point
+          :reader reference-point)))
+
+(defclass function-reference (reference)
+  ((arguments :type list
+              :initarg :function-arguments
+              :reader function-arguments)))
+
+(defclass package-reference (reference)
+  ((use :type list
+        :initarg :package-use
+        :reader package-use )
+   (export :type list
+           :initarg :package-export
+           :reader package-export)))
+
+(defclass class-reference (reference)
+  ((parents :type list
+            :initarg :class-parents
+            :reader class-parents)
+   (attributes :type list
+               :initarg :class-attributes
+               :reader class-attributes)))
+
+(defclass variable-reference (reference)
+  ((value :initarg :variable-reference-value
+          :reader variable-reference-value)))
+
+
+(defclass misc-reference (reference)
+  ((custom-type :initarg :misc-custom-type
+                :type string
+                :reader misc-custom-type)))
+
+(defstruct capture-regex regex function)
+
+(defclass detective-search () ())
+
+(defclass search-regex (detective-search)
+  ((function-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :function-regex
+                   :writer set-function-regex
+                   :reader search-function-regex)
+   (package-regex :type (or capture-regex null)
+                  :initform nil
+                  :initarg :package-regex
+                  :writer set-package-regex
+                  :reader search-package-regex)
+   (class-regex :type (or capture-regex null)
+                :initform nil
+                :initarg :class-regex
+                :writer set-class-regex
+                :reader search-class-regex)
+   (variable-regex :type (or capture-regex null)
+                   :initform nil
+                   :initarg :variable-regex
+                   :writer set-variable-regex
+                   :reader search-variable-regex )
+   (misc-regex :type (or capture-regex null)
+               :initform nil
+               :initarg :misc-regex
+               :writer set-misc-regex
+               :reader search-misc-regex)))
+
+
+(defun buffer-references (buffer)
+  (buffer-value buffer 'references))
+
+(defun (setf buffer-references) (value buffer)
+  (setf (buffer-value buffer 'references) value))
+
+(defgeneric search-references (search-class))
+
+(defmethod search-references ((search search-regex))
+  (labels ((find-ref (regex class)
+             (with-point ((p (buffer-start-point (point-buffer (current-point)))))
+               (loop :for position = (search-forward-regexp p (capture-regex-regex regex))
+                     :while position
+                     :collect (funcall (capture-regex-function regex) 
+                                       (copy-point position)
```

Comment:
This will probably never be released as copied.
If the buffer is never updated while using the copied point, it seems better to use `:temporary`.

Reply:
Sorry what do you mean with `:temporary`? Like another way of having the point?

Reply:
https://github.com/lem-project/lem/blob/dd0e5bf1b9a89aa61a5e5fc2e91cc348d3c5a805/src/base/point.lisp#L20
Would this be helpful?

Reply:
Oh, so copy the point with the kind as `:temporary`?

Reply:
Yes (but if you then change the buffer, that point may be broken)

Reply:
Umm, so if the buffer content got deleted, that point may not exist anymore. So, the idea is that shouldn't happen because of the check if the buffer was changed. (did I understand it correctly?)